test/conformance/helpers/sources/source_crd_rbac_test_helper.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sources
+
+import (
+	"context"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	testlib "knative.dev/eventing/test/lib"
+	"knative.dev/pkg/apis/duck"
+)
+
+var clusterRoleName = "eventing-sources-source-observer"
+var clusterRoleLabel = map[string]string{
+	duck.SourceDuckVersionLabel: "true",
+}
+
+func SourceCRDRBACTestHelperWithComponentsTestRunner(
+	t *testing.T,
+	sourceTestRunner testlib.ComponentsTestRunner,
+	options ...testlib.SetupClientOption,
+) {
+
+	sourceTestRunner.RunTests(t, testlib.FeatureBasic, func(st *testing.T, source metav1.TypeMeta) {
+		client := testlib.Setup(st, true, options...)
+		defer testlib.TearDown(client)
+
+		// From spec:
+		// Each source MUST have the following:
+		// kind: ClusterRole
+		// apiVersion: rbac.authorization.k8s.io/v1
+		// metadata:
+		//   name: foos-source-observer
+		//   labels:
+		//     duck.knative.dev/source: "true"
+		// rules:
+		//   - apiGroups:
+		//       - example.com
+		//     resources:
+		//       - foos
+		//     verbs:
+		//       - get
+		//       - list
+		//       - watch
+		st.Run("Source CRD has source observer cluster role", func(t *testing.T) {
+			ValidateRBAC(st, client, source)
+		})
+
+	})
+}
+
+func ValidateRBAC(st *testing.T, client *testlib.Client, object metav1.TypeMeta) {
+	labelSelector := &metav1.LabelSelector{
+		MatchLabels: clusterRoleLabel,
+	}
+
+	sourcePluralName := getSourcePluralName(client, object)
+
+	if !clusterRoleMeetsSpecs(client, labelSelector, sourcePluralName) {
+		//CRD Spec says new sources MUST include a ClusterRole as part of installing themselves into a cluster - so can't enforce it. Nothing to do here
+		//client.T.Fatalf("can't find source observer cluster role for CRD %q", object)
+		client.T.Logf("can't find source observer cluster role for CRD %q", object)
```

Comment:
Should this be an error or warning?

Reply:
Warning, since the spec says MUST instead of SHOULD, if that's ok? So I'll change it to warning.